olive/cache.py

```diff
@@ -32,127 +37,243 @@ class CacheSubDirs:
     @classmethod
     def from_cache_dir(cls, cache_dir: Path) -> "CacheSubDirs":
         return cls(
-            models=cache_dir / "models",
+            cache_dir=cache_dir,
             runs=cache_dir / "runs",
             evaluations=cache_dir / "evaluations",
             resources=cache_dir / "resources",
             mlflow=cache_dir / "mlflow",
         )
 
 
+class CacheConfig(ConfigBase):
+    cache_dir: Union[str, Path, List[str]] = DEFAULT_CACHE_DIR
+    clean_cache: bool = False
+    clean_evaluation_cache: bool = False
+    account_name: str = None
+    container_name: str = None
+    enable_shared_cache: bool = False
+    update_shared_cache: bool = True
+
+    @validator("account_name", pre=True)
+    def validate_account_name(cls, v, values):
+        if v:
+            return v
+        match = cls._get_shared_cache_match(values.get("cache_dir"))
+        return match.group(1) if match else None
+
+    @validator("container_name", pre=True)
+    def validate_container_name(cls, v, values):
+        if v:
+            return v
+        match = cls._get_shared_cache_match(values.get("cache_dir"))
+        return match.group(2) if match else None
+
+    @root_validator(pre=True)
+    def validate_enable_shared_cache(cls, values):
+        if values.get("account_name") and values.get("container_name"):
+            values["enable_shared_cache"] = True
+        elif values.get("enable_shared_cache"):
+            values["account_name"] = values.get("account_name") or "olivepublicmodels"
+            values["container_name"] = values.get("container_name") or "olivecachemodels"
+        else:
+            values["enable_shared_cache"] = False
+        return values
+
+    @staticmethod
+    def _get_shared_cache_match(cache_dir):
+        cache_dirs = [cache_dir] if not isinstance(cache_dir, list) else cache_dir
+        pattern = r"https://([^.]+)\.blob\.core\.windows\.net/([^/]+)"
+        for cache in cache_dirs:
+            match = re.match(pattern, str(cache))
+
+            if match:
+                return match
+        return None
+
+    def create_cache(self, workflow_id: str = DEFAULT_WORKFLOW_ID) -> "OliveCache":
+        self.cache_dir = Path(self.cache_dir) / workflow_id
+        return OliveCache(self)
+
+
 class OliveCache:
-    def __init__(
-        self,
-        cache_dir: Union[str, Path],
-        clean_cache: bool = False,
-        clean_evaluation_cache: bool = False,
-    ):
-        self.cache_dir = Path(cache_dir).resolve()
-        logger.info("Using cache directory: %s", self.cache_dir)
-        self.dirs = CacheSubDirs.from_cache_dir(self.cache_dir)
+    def __init__(self, cache_config: Union[CacheConfig, Dict]):
+        cache_config = validate_config(cache_config, CacheConfig)
+        cache_dir = Path(cache_config.cache_dir).resolve()
```

Comment:
Assuming `cache_dir` is a str or path. What about List[str]?

Reply:
Will fix this.

Reply:
fixed